File: crates/parity-proto/src/observation.rs
//! THE LAYERED OBSERVATION.
//!
//! Nothing here is normalized at capture. Canonicalization is a **projection applied
//! at compare time** (see `project.rs`). The trace on disk keeps every raw timestamp,
//! TTL, native error string and proto field; the comparator decides what to look at,
//! and a ledger claim may *widen* that view to compare a field the default hides.
//!
//! That ordering is the whole answer to "how do you canonicalize without destroying
//! the differences you are hunting?" — **you don't canonicalize the evidence, you
//! canonicalize the question.** Evidence discarded at capture is gone; a projection
//! can always be widened after the fact.
//!
//! Three layers, and each is diffed differently:
//!
//! | layer    | content                                        | diffed by default?              |
//! |----------|------------------------------------------------|---------------------------------|
//! | `class`  | closed vocabulary, from the kvproto TiKV sent  | **always** — the claim surface  |
//! | `proto`  | the `kvrpcpb` message as JSON                  | presence + type only; fields opt-in |
//! | `native` | the client's OWN taxonomy, verbatim            | **never** — but always *printed* |
//!
//! `native` is why this papers over nothing. Finding 1 IS a native-taxonomy bug:
//! `check_txn_status` matches `Error::ExtractedErrors` but the plan shape only ever
//! delivers `Error::MultipleKeyErrors`. Both are `Class::TxnNotFound` — the same
//! *fact* — so a class-only diff would show them as identical and the bug would be
//! invisible. Keeping `native.type` means a claim can opt into diffing precisely the
//! distinction that IS the defect.

use serde::Deserialize;
use serde::Serialize;

use crate::class::Class;

/// One command's result, as the driver saw it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Observation {
    /// L2 — the parity claim surface.
    #[serde(flatten)]
    pub class: Class,

    /// The value a read returned. Bytes, always base64 — never utf-8-lossy.
    /// (`gate::a2` exists precisely because values must stay opaque bytes; a lossy
    /// round-trip through a String would silently pass a test whose entire point is
    /// that it must not.)
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub value: Option<Bytes>,

    /// Locks observed by `scan_locks`. THE observation for findings 1 and 2, and the
    /// one place both clients can be asked the same question about durable residue.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub locks: Vec<LockObs>,

    /// Transaction timestamps. Recorded, never compared (see `project.rs`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub start_ts: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub commit_ts: Option<u64>,

    /// L1 — the `kvrpcpb` message, as canonical JSON, when the error carried one.
    ///
    /// Presence and message type are compared by default; field values are opt-in.
    /// That default is deliberate: it cheaply catches "Rust surfaced no proto at all,
    /// Go surfaced `AlreadyExist`" — the `insert`-on-existing-key divergence, where
    /// Rust's `DuplicateKeyInsertion` is a CLIENT-SIDE buffer check with no server
    /// round-trip and Go's `ErrKeyExist` comes back from prewrite — without drowning
    /// every diff in raw timestamps.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub proto: Option<serde_json::Value>,

    /// L0 — the client's own taxonomy, verbatim. NEVER diffed by default; ALWAYS
    /// printed in a divergence report, so a human can overturn any mapping call.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub native: Option<NativeObs>,
}

impl Observation {
    pub fn new(class: Class) -> Self {
        Self {
            class,
            value: None,
            locks: Vec::new(),
            start_ts: None,
            commit_ts: None,
            proto: None,
            native: None,
        }
    }

    pub fn ok() -> Self {
        Self::new(Class::Ok)
    }

    /// The driver could not do this at all, because the client cannot.
    pub fn unsupported(detail: impl Into<String>) -> Self {
        Self::new(Class::Unsupported {
            detail: detail.into(),
        })
    }

    /// The harness broke. Inadmissible; never a divergence.
    pub fn driver_error(detail: impl Into<String>) -> Self {
        Self::new(Class::DriverError {
            detail: detail.into(),
        })
    }

    pub fn with_value(mut self, v: Option<Vec<u8>>) -> Self {
        self.value = v.map(Bytes::from);
        self
    }

    pub fn with_native(mut self, native: NativeObs) -> Self {
        self.native = Some(native);
        self
    }

    pub fn with_proto(mut self, proto: serde_json::Value) -> Self {
        self.proto = Some(proto);
        self
    }

    pub fn with_locks(mut self, locks: Vec<LockObs>) -> Self {
        self.locks = locks;
        self
    }

    pub fn with_start_ts(mut self, ts: u64) -> Self {
        self.start_ts = Some(ts);
        self
    }

    pub fn with_commit_ts(mut self, ts: u64) -> Self {
        self.commit_ts = Some(ts);
        self
    }
}

/// Opaque bytes on the wire: base64, plus a `utf8` echo when printable.
///
/// Only `b64` is ever compared. `utf8` exists so a human reading a trace or a
/// divergence report can see `gate/d6/…/z-secondary` instead of a wall of base64 —
/// it is an affordance for the reader, never an input to the verdict.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Bytes {
    pub b64: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub utf8: Option<String>,
}

impl Bytes {
    pub fn as_slice(&self) -> Vec<u8> {
        b64_decode(&self.b64).unwrap_or_default()
    }
}

impl From<Vec<u8>> for Bytes {
    fn from(v: Vec<u8>) -> Self {
        Self {
            b64: b64_encode(&v),
            utf8: String::from_utf8(v.clone())
                .ok()
                .filter(|s| s.chars().all(|c| !c.is_control())),
        }
    }
}

/// A lock, as `scan_locks` reported it.
///
/// Deliberately the `kvrpcpb::LockInfo` fields both clients expose, and no more:
/// this is the symmetric observation that makes findings 1 and 2 diffable at all
/// (Rust: `TransactionClient::scan_locks -> Vec<kvrpcpb::LockInfo>`;
/// Go: `tikv.StoreProbe.ScanLocks -> []*txnlock.Lock`).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LockObs {
    pub key: Bytes,
    pub primary: Bytes,
    /// The `kvrpcpb::Op` enum BY NAME (`put` / `del` / `lock` / `pessimistic_lock`).
    /// Identical numbering in both generated protos, so the name is exact — and a
    /// name survives a proto renumbering that a bare integer would not.
    pub kind: String,
    /// Recorded, never compared: Rust's default TTL and Go's size-scaled TTL
    /// (`SetLockTTLByTimeAndSize`) genuinely differ, and that is a difference we
    /// choose not to claim parity on. `project.rs` buckets it to {zero, positive}.
    pub ttl_ms: u64,
    pub txn_start_ts: u64,
}

/// The client's own error taxonomy — the layer that keeps the mapping honest.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct NativeObs {
    /// "rust" | "go"
    pub lang: String,
    /// The client's own type name: `MultipleKeyErrors`, `*tikverr.ErrWriteConflict`.
    pub r#type: String,
    /// `Debug`/`Error()` output, verbatim.
    pub display: String,
}

impl NativeObs {
    pub fn new(lang: &str, r#type: impl Into<String>, display: impl Into<String>) -> Self {
        Self {
            lang: lang.to_owned(),
            r#type: r#type.into(),
            display: display.into(),
        }
    }
}

// ── base64, hand-rolled ──────────────────────────────────────────────────────
// A dependency-free alphabet. The harness already refuses `rand` in favour of a
// 6-line xorshift (`harness::deterministic_bytes`); a base64 crate for ~30 lines of
// table lookup is the same trade, and every dependency in the runner is one more
// thing that has to be true for a verdict to mean anything.

const ALPHABET: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

pub fn b64_encode(input: &[u8]) -> String {
    let mut out = String::with_capacity(input.len().div_ceil(3) * 4);
    for chunk in input.chunks(3) {
        let b = [
            chunk[0],
            *chunk.get(1).unwrap_or(&0),
            *chunk.get(2).unwrap_or(&0),
        ];
        let n = (u32::from(b[0]) << 16) | (u32::from(b[1]) << 8) | u32::from(b[2]);
        out.push(ALPHABET[(n >> 18) as usize & 63] as char);
        out.push(ALPHABET[(n >> 12) as usize & 63] as char);
        out.push(if chunk.len() > 1 {
            ALPHABET[(n >> 6) as usize & 63] as char
        } else {
            '='
        });
        out.push(if chunk.len() > 2 {
            ALPHABET[n as usize & 63] as char
        } else {
            '='
        });
    }
    out
}

pub fn b64_decode(input: &str) -> Option<Vec<u8>> {
    let mut buf = 0u32;
    let mut bits = 0u32;
    let mut out = Vec::new();
    for c in input.bytes() {
        if c == b'=' {
            break;
        }
        let v = ALPHABET.iter().position(|&a| a == c)? as u32;
        buf = (buf << 6) | v;
        bits += 6;
        if bits >= 8 {
            bits -= 8;
            out.push((buf >> bits) as u8);
        }
    }
    Some(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn base64_round_trips_arbitrary_bytes() {
        // Every length mod 3, plus the bytes that a utf-8 round-trip would destroy.
        for case in [
            vec![],
            vec![0x00],
            vec![0xff, 0xfe],
            vec![1, 2, 3],
            vec![0xde, 0xad, 0xbe, 0xef],
            (0u8..=255).collect::<Vec<_>>(),
        ] {
            let enc = b64_encode(&case);
            assert_eq!(b64_decode(&enc).unwrap(), case, "{enc}");
        }
    }

    #[test]
    fn bytes_keeps_non_utf8_values_as_bytes() {
        // CAS soundness (gate::a2) turns on values being byte-identical. A value that
        // is not valid utf-8 must survive; the `utf8` echo is only a reader affordance.
        let raw = vec![0xff, 0x00, 0xfe];
        let b = Bytes::from(raw.clone());
        assert_eq!(b.as_slice(), raw);
        assert!(b.utf8.is_none(), "invalid utf-8 must not be echoed");
    }

    #[test]
    fn bytes_echoes_printable_utf8_for_the_reader() {
        let b = Bytes::from(b"gate/d6/z-secondary".to_vec());
        assert_eq!(b.utf8.as_deref(), Some("gate/d6/z-secondary"));
        assert_eq!(b.as_slice(), b"gate/d6/z-secondary");
    }

    #[test]
    fn a_value_containing_a_newline_is_not_echoed() {
        // The trace is newline-delimited JSON at the transport layer; a control
        // character in the `utf8` echo is a footgun waiting for a careless reader.
        let b = Bytes::from(b"a\nb".to_vec());
        assert!(b.utf8.is_none());
    }
}

File: go/driver/observation.go
// THE GO HALF OF THE MAPPING TABLE.
//
// Its Rust twin is crates/rust-driver/src/mapping.rs. Between them they define what
// it MEANS for two clients to have done the same thing, so this is the one file where
// a lie could hide — and the four defenses against that are:
//
//  1. Nothing is normalized at capture. `class` is a projection of what the SERVER
//     said; the client's own error is carried verbatim in `native` and the kvproto in
//     `proto`. A mapping call can always be overturned from the evidence.
//  2. `native` is ALWAYS attached. If this table maps two different facts to one
//     class, the trace still contains the difference and a ledger claim can opt in to
//     diffing `native.type`. (Finding 1 lives exactly there.)
//  3. Every arm below cites the client-go source it mirrors, by file and line.
//  4. The catch-all is `internal`, which is INADMISSIBLE. An unmapped error can never
//     become `ok` or `write_conflict`; it becomes a run ledger-check refuses. A result
//     that MIGHT be a lie is worth less than no result at all.
//
// THE RULE: normalize presentation, never fact. Go returning ErrNotExist where Rust
// returns Ok(None) is one fact spelled two ways -> both `not_found`. Rust failing to
// resolve a lock where Go succeeds is TWO FACTS, and must stay two.
package main

import (
	"encoding/base64"
	"errors"
	"fmt"
	"regexp"
	"strings"
	"unicode"

	"github.com/pingcap/kvproto/pkg/kvrpcpb"
	tikverr "github.com/tikv/client-go/v2/error"
	"github.com/tikv/client-go/v2/txnkv/txnlock"
)

// Observation is one command's result. Mirrors parity-proto's Rust type exactly —
// by hand: there is no generated binding and no golden corpus (yet), so the two
// halves are kept in lockstep by review. A field added on one side without the other
// serializes as absent and the projection treats absent as a fact, so drift here is
// not cosmetic.
type Observation struct {
	Class string `json:"class"`

	// Set only for the classes that carry one (unsupported / driver_error / internal).
	Detail string `json:"detail,omitempty"`
	// Set only for `mixed`.
	Parts []string `json:"parts,omitempty"`

	Value *Bytes `json:"value,omitempty"`
	Locks []Lock `json:"locks,omitempty"`

	StartTS  uint64 `json:"start_ts,omitempty"`
	CommitTS uint64 `json:"commit_ts,omitempty"`

	// How many per-key errors this client surfaced. For client-go the answer is always
	// ONE: ExtractKeyErr (error/error.go:331) returns a single error even when several
	// keys failed. client-rust surfaces all of them.
	//
	// It must be emitted, not left absent, or a claim opting into `errors.count` would
	// compare Go's `<absent>` against Rust's "1" and diverge even when both clients
	// surfaced exactly one error — a false divergence manufactured by an unset field. And
	// "client-go collapses N errors into 1" is a real, statable parity fact; leaving the
	// field empty would make it indistinguishable from "we did not look".
	ErrorCount *int `json:"error_count,omitempty"`

	Proto  map[string]any `json:"proto,omitempty"`
	Native *Native        `json:"native,omitempty"`

	// The result of a raw checksum, when this client could compute one. EVIDENCE
	// ONLY, never projected: crc64_xor/total_bytes cover the key bytes — run prefix
	// included — so they legitimately differ between the two runs of one comparison.
	Checksum *ChecksumObs `json:"checksum,omitempty"`
}

// ChecksumObs mirrors parity_proto::ChecksumObs (kvrpcpb.RawChecksumResponse, verbatim).
type ChecksumObs struct {
	Crc64Xor   uint64 `json:"crc64_xor"`
	TotalKvs   uint64 `json:"total_kvs"`
	TotalBytes uint64 `json:"total_bytes"`
}

type Bytes struct {
	B64  string  `json:"b64"`
	UTF8 *string `json:"utf8,omitempty"`
}

type Lock struct {
	Key        Bytes  `json:"key"`
	Primary    Bytes  `json:"primary"`
	Kind       string `json:"kind"`
	TTLms      uint64 `json:"ttl_ms"`
	TxnStartTS uint64 `json:"txn_start_ts"`
}

type Native struct {
	Lang    string `json:"lang"`
	Type    string `json:"type"`
	Display string `json:"display"`
}

func mkBytes(raw []byte) Bytes {
	b := Bytes{B64: base64.StdEncoding.EncodeToString(raw)}
	if s := string(raw); isPrintable(s) {
		b.UTF8 = &s
	}
	return b
}

func isPrintable(s string) bool {
	for _, r := range s {
		if r == unicode.ReplacementChar || unicode.IsControl(r) {
			return false
		}
	}
	return true
}

func ok() *Observation                  { return &Observation{Class: "ok"} }
func notFound() *Observation            { return &Observation{Class: "not_found"} }
func unsupported(d string) *Observation { return &Observation{Class: "unsupported", Detail: d} }
func driverError(d string) *Observation { return &Observation{Class: "driver_error", Detail: d} }

// lockKind renders kvrpcpb.Op BY NAME. Both clients generate this enum from the same
// proto, so the name is exact — and a name survives a renumbering that a bare integer
// would not.
// The mapping must be TOTAL over kvrpcpb.Op, and it must agree with the Rust driver's
// `lock_kind` arm for arm. Both clients generate this enum from the same proto, so any
// operation one of them names and the other renders as a bare `op_N` is a divergence the
// HARNESS invented — the clients returned the identical protobuf value.
//
// Op_Rollback was exactly that: Rust said "rollback", Go said "op_3".
func lockKind(op kvrpcpb.Op) string {
	switch op {
	case kvrpcpb.Op_Put:
		return "put"
	case kvrpcpb.Op_Del:
		return "del"
	case kvrpcpb.Op_Lock:
		return "lock"
	case kvrpcpb.Op_Rollback:
		return "rollback"
	case kvrpcpb.Op_Insert:
		return "insert"
	case kvrpcpb.Op_PessimisticLock:
		return "pessimistic_lock"
	case kvrpcpb.Op_CheckNotExists:
		return "check_not_exists"
	default:
		// Deliberately still reachable: if the proto grows an operation, BOTH drivers
		// render it as `op_N` and agree, rather than one of them guessing at a name.
		return fmt.Sprintf("op_%d", int32(op))
	}
}

func mkLocks(locks []*txnlock.Lock) []Lock {
	out := make([]Lock, 0, len(locks))
	for _, l := range locks {
		out = append(out, Lock{
			Key:        mkBytes(l.Key),
			Primary:    mkBytes(l.Primary),
			Kind:       lockKind(l.LockType),
			TTLms:      l.TTL,
			TxnStartTS: l.TxnID,
		})
	}
	return out
}

// txnNotFoundRe matches client-go's UNTYPED TxnNotFound.
//
// THE ONE UGLY ARM, and it is unavoidable. Every other KeyError becomes a typed error
// in ExtractKeyErr, but TxnNotFound does not:
//
//	client-go error/error.go:363
//	    if keyErr.TxnNotFound != nil {
//	        err := errors.Errorf("txn %d not found", keyErr.TxnNotFound.StartTs)
//	        return err
//	    }
//
// There is no typed error and no wrapped proto to match on, so the only handle is the
// message text. That is fragile, and it is precisely why it is confined to ONE place
// with this comment: if client-go ever rewords it, this stops matching and the arm
// falls through to `internal` — INADMISSIBLE, a loud failure, not a silent
// misclassification. The failure mode is correct even when the match is wrong.
//
// (This asymmetry is itself worth noting upstream: TxnNotFound is the one KeyError
// client-go leaves untyped, and it happens to be the one finding 1 turns on.)
var txnNotFoundRe = regexp.MustCompile(`^txn \d+ not found`)

// classify maps a client-go error onto the canonical vocabulary.
//
// Order matters: the typed errors first, the string-matched wart last, and an explicit
// `internal` for anything unrecognized.
func classify(err error) *Observation {
	if err == nil {
		return ok()
	}

	// ── not found: a NORMAL OUTCOME, not an error ────────────────────────────
	// Checked FIRST, and deliberately WITHOUT a `native` block.
	//
	// client-go signals "the key is absent" with a sentinel error (error/error.go:60,
	// ErrNotExist); client-rust signals it with Ok(None). Same fact, different calling
	// convention. If we attached Go's native error type here, then EVERY not_found
	// comparison would carry `native.type = *errors.fundamental` on the Go side and
	// `<absent>` on the Rust side, and any claim that opts into diffing native.type
	// would see a PERMANENT divergence that is not a gap at all — it is a fact about
	// Go's calling convention, not about TiKV.
	//
	// This is the boundary the mapping rule draws: normalize PRESENTATION (how a
	// language spells "absent"), never FACT. `native` exists to preserve a client's
	// taxonomy OF AN ERROR; a successful absent-key read has no error to have a
	// taxonomy of.
	if tikverr.IsErrNotFound(err) {
		return notFound()
	}

	// client-go surfaces exactly one error, however many keys failed: ExtractKeyErr
	// returns a single error. Stating that explicitly is what lets a claim compare
	// cardinality against client-rust, which surfaces all of them.
	one := 1
	obs := &Observation{
		ErrorCount: &one,
		Native: &Native{
			Lang:    "go",
			Type:    fmt.Sprintf("%T", errors.Unwrap(err)),
			Display: err.Error(),
		},
	}
	// %T of the unwrapped error is often more informative, but can be nil for a
	// leaf error; fall back to the error's own type.
	if obs.Native.Type == "<nil>" {
		obs.Native.Type = fmt.Sprintf("%T", err)
	}

	// ── write conflict — the lost-race signal ────────────────────────────────
	// client-go error/error.go:166. Carries the kvrpcpb.WriteConflict proto.
	var wc *tikverr.ErrWriteConflict
	if errors.As(err, &wc) {
		obs.Class = "write_conflict"
		obs.Proto = map[string]any{
			"key_error": map[string]any{
				"conflict": map[string]any{
					"start_ts":           wc.StartTs,
					"conflict_ts":        wc.ConflictTs,
					"conflict_commit_ts": wc.ConflictCommitTs,
					"reason":             wc.Reason.String(),
				},
			},
		}
		return obs
	}

	// ── key exists ───────────────────────────────────────────────────────────
	// client-go error/error.go — ErrKeyExist wraps kvrpcpb.AlreadyExist, i.e. it
	// comes back from a real PREWRITE. Rust's DuplicateKeyInsertion is a CLIENT-SIDE
	// buffer check with no proto at all. Same class, and the `proto.present`
	// divergence is what exposes that they are not the same event.
	var ke *tikverr.ErrKeyExist
	if errors.As(err, &ke) {
		obs.Class = "key_exists"
		obs.Proto = map[string]any{
			"key_error": map[string]any{"already_exist": map[string]any{}},
		}
		return obs
	}

	var dl *tikverr.ErrDeadlock
	if errors.As(err, &dl) {
		obs.Class = "deadlock"
		obs.Proto = map[string]any{
			"key_error": map[string]any{"deadlock": map[string]any{}},
		}
		return obs
	}

	var af *tikverr.ErrAssertionFailed
	if errors.As(err, &af) {
		obs.Class = "assertion_failed"
		obs.Proto = map[string]any{
			"key_error": map[string]any{"assertion_failed": map[string]any{}},
		}
		return obs
	}

	var rt *tikverr.ErrRetryable
	if errors.As(err, &rt) {
		obs.Class = "retryable"
		return obs
	}

	// ── undetermined ─────────────────────────────────────────────────────────
	// The commit MAY have landed. Never roll back on this — doing so could tear a
	// committed batch. Both clients must surface it distinguishably.
	if tikverr.IsErrorUndetermined(err) {
		obs.Class = "undetermined"
		return obs
	}

	// ── txn not found — the string-matched wart (see txnNotFoundRe) ──────────
	if txnNotFoundRe.MatchString(err.Error()) {
		obs.Class = "txn_not_found"
		obs.Proto = map[string]any{
			"key_error": map[string]any{"txn_not_found": map[string]any{}},
		}
		return obs
	}

	// ── still-locked ─────────────────────────────────────────────────────────
	// A lock the client did NOT resolve. For client-go this is rare by construction:
	// its LockResolver resolves on read. When Rust surfaces this and Go does not,
	// that IS the finding.
	if strings.Contains(err.Error(), "key is locked") {
		obs.Class = "unresolved_lock"
		obs.Proto = map[string]any{
			"key_error": map[string]any{"locked": map[string]any{}},
		}
		return obs
	}

	// ── region / pd / rpc ────────────────────────────────────────────────────
	// A region error is a CLUSTER event, not client behaviour: the trace becomes
	// inadmissible and the runner retries, rather than reporting a divergence that
	// is really about the cluster.
	if strings.Contains(err.Error(), "region") && strings.Contains(err.Error(), "epoch") {
		obs.Class = "region_error"
		return obs
	}

	// ── THE CATCH-ALL: inadmissible, on purpose ──────────────────────────────
	// This is the whole defense. An error nobody mapped must NEVER be quietly
	// rendered as some plausible class — it becomes a run that cannot settle a
	// claim, and someone has to extend the taxonomy deliberately.
	obs.Class = "internal"
	obs.Detail = fmt.Sprintf("unmapped client-go error (%s): %s", obs.Native.Type, err.Error())
	return obs
}
